### ISSUE TITLE
Running tensorflow on GPU is far slower than on CPU

### ISSUE THREAD

**Issue by forReason** [tensorflow/tensorflow#31654]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10 1809 & Windows Server 2016
- TensorFlow installed from (source or binary): pip install tensorflow-gpu==2.0.0-beta1, as well as tensorflow-gpu, compared to tensorflow & tensorflow==2.0.0-beta1
- Python version: 3.6
- Bazel version (if compiling from source): -
- GCC/Compiler version (if compiling from source): - 
- CUDA/cuDNN version: Cuda 10 & cuDNN 7.6.2 for cuda 10.0

### Current behavior:
Im getting a 50%+ performance loss with GPU!
In the below example, the CPU version is even training way faster on a bigger model with slightly bigger epochs.
![time](https://user-images.githubusercontent.com/12736950/63144316-9daf1a00-bff2-11e9-9b39-036fd9291008.png)

Im training on 2 different systems:
My server, without a GPU:

Intel Core i5 6500T (4x@2.5 ghz) Notebook processor
Network attached storage for training data and output
16 Gb DDR 4 Ram
My Desktop PC:

Intel core i7 3770k (4x3,5-4 ghz)
Nvidia GTX 970 @4GB
32gb ddr3 ram
Training Data on local SSD, output to NAS

Now interestingly 3000 epochs, 100000 records each, takes roughly 3h on the server using TF 1.14
The same on my Desktop with GPU takes 8h with TF 2.0
It sits with full Video Ram but at 3% graphical processor use.
The CPU is sometimes at 30% use with tensorflow GPU but 100%  at any time with any CPU build.  
The harddisk is utilized with a whopping 0%.
![system utilisation](https://user-images.githubusercontent.com/12736950/63122624-7f6ffc80-bfa7-11e9-81ee-2b76e4bc5d99.png)

### Expected behavior:
Tensorflow-GPU trains faster than Tensorflow CPU

### Code to reproduce the issue:
My model is a fairly simple keras sequential lstm:
```
for learningrate in learningrates:
	for layerdensity in layerdensitys:
		for layer in amount_of_layers:
			################################
			# generate model               #
			################################
			modelname = f"{layer}-layer_{layerdensity}-nodes_selu-adam_{learningrate}-learningrate_{records_per_epoch}-epochsize_{appendix}"
			model = keras.Sequential()
			model.add(Dense(layerdensity, activation=tf.nn.selu, input_dim=15))
			for i in range(layer-1):
				model.add(Dense(layerdensity, activation=tf.nn.selu))
			model.add(Dense(9,activation=tf.nn.softmax, name = "Output"))
			# Compile
			optimizer = tf.keras.optimizers.Adam(lr=learningrate)
			model.compile(
				optimizer=optimizer,
				loss='sparse_categorical_crossentropy',
				metrics=['accuracy'])
			model.summary()
			tensorboard = TensorBoard(log_dir="\\\\drg-fs01\\BigData\\Projects\\Notebooks\\PokerBot\\log\\" + modelname,
				histogram_freq = 100, write_graph = False)
			#cp_callback = tf.keras.callbacks.ModelCheckpoint("\\\\drg-fs01\\BigData\\Projects\\Notebooks\\PokerBot\\checkpoints\\" + modelname, verbose=0)
			################################
			# train model                  #
			################################
			model.fit(trainSet, 
				epochs = epochs, 
				steps_per_epoch = trainSteps, 
				shuffle = True, 
				validation_data = testSet, 
				validation_steps = testSteps, 
				validation_freq = int(epochs/maxTestEpochs),
				verbose = verbose, 
				callbacks = [tensorboard])#,cp_callback])
			model.save(basePath+'saved_models/' + modelname + '.h5')
```
I am having lots of test data which does not fit in memory so I use interleaved datasets:
```
# dataset modeler
def modelDataset(sourcepath, badgesize, repeat = False, repetitions = 10):
    #get all files
    list = os.listdir(sourcepath)
    pathfiles = [sourcepath+x for x in list]
    
    #get metrics
    rows_per_file = count_lines(sourcepath+"0.csv")
    number_of_files = len(list)
    total_rows = (rows_per_file * number_of_files)
    print(f"records: {total_rows}")
    # get number of steps per Epoch
    steps_per_epoch = int(rows_per_file / badgesize) # 2000 badges per epoch
    epochs = number_of_files
    if badgesize == 1:
        epochs = 1 
    print(f"number epochs: {epochs}")
    # model interleaved dataset
    dataset = (tf.data.Dataset.from_tensor_slices(pathfiles).interleave(lambda x:
        tf.data.TextLineDataset(x).map(parse_csv, num_parallel_calls=4),
        cycle_length=4, block_length=16))
    dataset.columns = CSV_COLUMNS
    
    if badgesize != 1:
        dataset = dataset.shuffle(buffer_size=badgesize)
    if repeat:
        dataset = dataset.repeat(repetitions)
        epochs = epochs * repetitions
    dataset = dataset.batch(badgesize)
    dataset = dataset.prefetch(2)  # prefetch one batch
    return dataset, steps_per_epoch, epochs, badgesize
```

**Comment by gadagashwini** [comment]
@forReason Provide us the full code snippet to reproduce the reported issue. It will indeed help us to move faster. Thanks!

**Comment by forReason** [comment]
I have been able to increase the speed by upping the Batchsize to 128. 
Still taskmanager shows a low gpu utilisation but the cpu utilisation increased.
![2019-08-16 13_11_51-Task Manager](https://user-images.githubusercontent.com/12736950/63164174-ecc37200-c027-11e9-9860-563e10fd1cdd.png)

here is my full script trainmodel.py:

```
from __future__ import absolute_import, division, print_function, unicode_literals

from sklearn.model_selection import train_test_split

# TensorFlow and tf.keras
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.initializers import glorot_uniform
from tensorflow import feature_column
from tensorflow.keras import layers
from tensorflow.keras.callbacks import TensorBoard
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense, Activation, Embedding, Flatten, LeakyReLU, BatchNormalization, Dropout
import tensorflow_datasets as tfds

# Helper libraries
import numpy as np
import time
import os
from tqdm import tqdm #progressbars

print(tf.__version__)

# count lines
def count_lines(path):
    rownumber = 0
    with open(path, "r",encoding="utf-8",errors='ignore') as f:
        rownumber =  sum(bl.count("\n") for bl in blocks(f))
    return rownumber

def blocks(files, size=65536):
    while True:
        b = files.read(size)
        if not b: break
        yield b

basePath = "//[fileshare]/BigData/Projects/Notebooks/Trainmodel/"
localPath = "C:/temp/Trainmodel/"
                                                                                  
valPath = localPath+"val/"
trainPath = localPath+"train/"
testPath = localPath+"test/"

# csv to tensor parser
def parse_csv(line):
        parsed_line = tf.io.decode_csv(line, [[0.], [0.], [0.],[0.], [0.], [0.], [0.],[0.], [0.], [0.], [0.],[0.], [0.], [0.], [0.], [0]])
        label = parsed_line[-1:]
        del parsed_line[-1]
        features = parsed_line
        return tf.stack(features),label

# dataset modeler
def modelDataset(sourcepath, badgesize, repeat = False, repetitions = 10):
    #get all files
    list = os.listdir(sourcepath)
    pathfiles = [sourcepath+x for x in list]
    
    #get metrics
    rows_per_file = count_lines(sourcepath+"0.csv")
    number_of_files = len(list)
    total_rows = (rows_per_file * number_of_files)
    print(f"records: {total_rows}")
    # get number of steps per Epoch
    steps_per_epoch = int(rows_per_file / badgesize) # 2000 badges per epoch
    epochs = number_of_files
    if badgesize == 1:
        epochs = 1 
    print(f"number epochs: {epochs}")
    # model interleaved dataset
    dataset = (tf.data.Dataset.from_tensor_slices(pathfiles).interleave(lambda x:
        tf.data.TextLineDataset(x).map(parse_csv, num_parallel_calls=4),
        cycle_length=4, block_length=16))
    #dataset.columns = CSV_COLUMNS
    
    if badgesize != 1:
        dataset = dataset.shuffle(buffer_size=badgesize)
    if repeat:
        dataset = dataset.repeat(repetitions)
        epochs = epochs * repetitions
    dataset = dataset.batch(badgesize)
    dataset = dataset.prefetch(2)  # prefetch one batch
    return dataset, steps_per_epoch, epochs, badgesize

# load interleaved dataset
trainSet, trainSteps, maxTrainEpochs, trainBadgeSize = modelDataset(trainPath, 256, True)
print (f"max number of epochs: {maxTrainEpochs}")
testSet, testSteps, maxTestEpochs, testBadgeSize = modelDataset(testPath, 256,True,4)
valSet, valSteps, maxValEpochs, valBadgeSize = modelDataset(valPath, 1)

# compile model
learningrates = [0.0002]
layerdensitys = [40]
amount_of_layers = [5]
appendix = "GPUtest"
epochs = maxTrainEpochs
records_per_epoch = trainSteps * trainBadgeSize
verbose = 0

################################
# train model                  #
################################
modelname = "unspecified"
for learningrate in learningrates:
	for layerdensity in layerdensitys:
		for layer in amount_of_layers:
			################################
			# generate model               #
			################################
			modelname = f"{layer}-layer_{layerdensity}-nodes_selu-adam_{learningrate}-learningrate_{records_per_epoch}-epochsize_{appendix}"
			model = keras.Sequential()
			model.add(Dense(layerdensity, activation=tf.nn.selu, input_dim=15))
			for i in range(layer-1):
				model.add(Dense(layerdensity, activation=tf.nn.selu))
			model.add(Dense(9,activation=tf.nn.softmax, name = "Output"))
			# Compile
			optimizer = tf.keras.optimizers.Adam(lr=learningrate)
			model.compile(
				optimizer=optimizer,
				loss='sparse_categorical_crossentropy',
				metrics=['accuracy'])
			model.summary()
			tensorboard = TensorBoard(log_dir="\\\\[fileshare]\\BigData\\Projects\\Notebooks\\Trainmodel\\log\\" + modelname,
				histogram_freq = 100, write_graph = False)
            #cp_callback = tf.keras.callbacks.ModelCheckpoint("\\\\[Fileshare]\\BigData\\Projects\\Notebooks\\Trainmodel\\checkpoints\\" + modelname, verbose=0)
            ################################
            # train model                  #
            ################################
			model.fit(trainSet, 
				epochs = epochs, 
				steps_per_epoch = trainSteps, 
				shuffle = True, 
				validation_data = testSet, 
				validation_steps = testSteps, 
				validation_freq = int(epochs/maxTestEpochs),
				verbose = verbose, 
				callbacks = [tensorboard])#,cp_callback])
			model.save(basePath+'saved_models/' + modelname + '.h5')
```

**Comment by esslushy** [comment]
I have also faced this issue when training a GAN on gpu, but I do see sparks up to .3% utilization when placement is default. Manually placing had no effect, but I thought it may be because slow transfer/read from cpu to gpu, but this makes it seem like something else. Thanks for doing the research on this @forReason

**Comment by forReason** [comment]
@esslushy Really nothing jumps to my mind. First, I thought it might be the IO on network so I pulled 50GB of data locally *grr*. didnt help. 
Having Drive and Network at Literally 0% is a headscratcher to me.
Next, I thought it might be the parse function:
```
# csv to tensor parser
def parse_csv(line):
        parsed_line = tf.io.decode_csv(line, [[0.], [0.], [0.],[0.], [0.], [0.], [0.],[0.], [0.], [0.], [0.],[0.], [0.], [0.], [0.], [0]])
        label = parsed_line[-1:]
        del parsed_line[-1]
        features = parsed_line
        return tf.stack(features),label
```

But why would it be faster on a slower cpu where it has to parse and train on the same chip?
When preparing the data with my own c# Program, I can shuffle, augment data and write to disk roughly 1ep/3secs (100'000 records)

but when training, I can get 1 step in 2.45s

**Comment by michaelmunie** [comment]
I think that your CPU performance is better than GPU performance because you have a relatively small model.  You should try a test run with 
`amount_of_layers = [5, 50, 150]`
And see if GPU is still performing slower than CPU.


As a poster on S.O. says, "the overhead of invoking GPU kernels, and copying data to and from GPU, is very high. For operations on models with very little parameters it is not worth of using GPU"
[Related StackOverflow Post](https://stackoverflow.com/questions/55749899/training-a-simple-model-in-tensorflow-gpu-slower-than-cpu)

**Comment by sanjoy** [comment]
Additionally, you can also try to use the [TensorBoard profiler](https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras) to see if that points to any red flags.

**Comment by forReason** [comment]
Hello Guys,

Issue seems to be fixed with TF 2.1 for me as well as Graphics card swap to RTX 2080.
Still many thanks for your assistance.

**Comment by vitaly-t** [comment]
I've just spent half a day, trying to get through the flimsy installation of GPU support for Windows 10, and when I finally got it all working, I got my GTX 780 performing about 10 times worse than my CPU. That's bad.

**Comment by arruda** [comment]
> Hello Guys,
> 
> Issue seems to be fixed with TF 2.1 for me as well as Graphics card swap to RTX 2080.
> Still many thanks for your assistance.

So the only solution is updating TF? that's bad since I have a older GPU and as far as I remember the newer TF are not compatible with older cuda versions... And what exactly was the problem, I'm a bit confused with your solution, the solution was to buy a new GPU? or update TF? or only by doing both?
Thanks =)

**Comment by jeugregg** [comment]
> I've just spent half a day, trying to get through the flimsy installation of GPU support for Windows 10, and when I finally got it all working, I got my GTX 780 performing about 10 times worse than my CPU. That's bad.

Did you solve your performance problem ? 
I have the same problem : GPU : GeForce GTX 1080 8Go, CPU : Intel Xeon E5-2667 v3 @ 3.20GHz RAM 12Mo...
I don't understand if I really use the GPU...

**Comment by forReason** [comment]
> > Hello Guys,
> > Issue seems to be fixed with TF 2.1 for me as well as Graphics card swap to RTX 2080.
> > Still many thanks for your assistance.
> 
> So the only solution is updating TF? that's bad since I have a older GPU and as far as I remember the newer TF are not compatible with older cuda versions... And what exactly was the problem, I'm a bit confused with your solution, the solution was to buy a new GPU? or update TF? or only by doing both?
> Thanks =)

Hey Arruda,

unfortunately I can not specify as I upgraded both at the same time.

Apparently, I have to upgrade my cpu now since after upgrading GPU now my CPU is the Bottleneck... 7 Years old computer... But its still a lot faster.

II guess at this point, that more powerful hardware = faster processing no matter if there is a bug or not.

**Comment by MikhailTroshin** [comment]
[Here](https://medium.com/lsc-psd/tensorflow-2-1-doesnt-seem-to-see-my-gpu-even-though-cuda-10-1-with-solution-7b44297843a) you can find a very usefull tutorial. I faced the same problem, and that post really helped. The issue was in cudnn- and cuda- conda driver versions incompatibility.
Hope, this will help

**Comment by Jesper-Hustad** [comment]
Changing my batch size from `10` to `2048` increased performance from about 6 minutes to 15 seconds.
